pkg/rollup/rollup.go

```diff
@@ -14,93 +14,92 @@ import (
-// matches the provided newResolutions.
-//
-// Example: If the DB already contains metric rollups for `short` and `long`, and in dataset-config,
-// connector sees `very_short` and `long`, then EnsureRollupWith will remove the `short` downsampled data
-// and create `very_short`, while not touching `long`.
-func EnsureRollupWith(conn *pgx.Conn, newResolutions []DownsampleResolution) error {
+// Sync updates the rollups in the DB in accordance with the given newResolutions. It handles:
+// 1. Creating of new rollups
+// 2. Deletion of rollups that have `delete: true`
+// 3. Update retention duration of rollups that have same label name but different retention duration
+func Sync(conn *pgx.Conn, newResolutions Resolution) error {
 	rows, err := conn.Query(context.Background(), "SELECT name, resolution, retention FROM _prom_catalog.rollup")
 	if err != nil {
 		return fmt.Errorf("querying existing resolutions: %w", err)
 	}
 	defer rows.Close()
 
-	var existingResolutions []DownsampleResolution
+	existingResolutions := make(Resolution)
 	for rows.Next() {
 		var lName string
 		var resolution, retention time.Duration
 		if err := rows.Scan(&lName, &resolution, &retention); err != nil {
 			return fmt.Errorf("error scanning output rows for existing resolutions: %w", err)
 		}
-		existingResolutions = append(existingResolutions, DownsampleResolution{Label: lName, Resolution: day.Duration(resolution), Retention: day.Duration(retention)})
+		existingResolutions[Label(lName)] = Info{Resolution: day.Duration(resolution), Retention: day.Duration(retention)}
 	}
 
-	// Determine which resolutions need to be created and deleted from the DB.
-	pendingCreation := diff(newResolutions, existingResolutions)
-	pendingDeletion := diff(existingResolutions, newResolutions)
+	if err := updateExistingRollups(conn, existingResolutions, newResolutions); err != nil {
+		return fmt.Errorf("update existing rollups: %w", err)
+	}
 
 	// Delete rollups that are no longer required.
-	if err = deleteRollups(conn, pendingDeletion); err != nil {
+	if err = deleteRollups(conn, existingResolutions, newResolutions); err != nil {
 		return fmt.Errorf("delete rollups: %w", err)
 	}
 
 	// Create new rollups.
-	if err = createRollups(conn, pendingCreation); err != nil {
+	if err = createRollups(conn, existingResolutions, newResolutions); err != nil {
 		return fmt.Errorf("create rollups: %w", err)
 	}
 	return nil
 }
 
-func createRollups(conn *pgx.Conn, res []DownsampleResolution) error {
-	for _, r := range res {
-		_, err := conn.Exec(context.Background(), "CALL _prom_catalog.create_rollup($1, $2, $3)", r.Label, time.Duration(r.Resolution), time.Duration(r.Retention))
-		if err != nil {
-			return fmt.Errorf("error creating rollup for %s: %w", r.Label, err)
+// updateExistingRollups updates the existing rollups retention if the new resolutions with a same name has
+// different retention duration.
+func updateExistingRollups(conn *pgx.Conn, existingRes, newRes Resolution) error {
```

Comment:
What is happening if user tweaks resolution for the existing rollup? I guess we should handle that in some way. 
Also we need to fix naming here. `Resolution` type and argument names are misleading since we are passing in Rollup definitions.

Reply:
If the user tweaks resolutions, we cannot do anything and hence should error. I will add this in the code.